contents/linux/shell.md

```diff
@@ -0,0 +1,180 @@
+
+```bash
+./calculate < DataBundle > result
+./verify < result
+```
+Notice in this case, `calculate` uses the data in `DataBundle` as input, and output the results to the file `result`. Then, `result` is taken as input to `verify`, and the final verification result is printed on screen.
+
+Furthermore, if the code we write is buggy, it throws exception and error messages are directed to error output, then our result file will be empty! To handle it properly, we can also redirect error output to a file `errors`, and verify it is empty(such that no error occurs during execution) before we execute `verify`.
+
+```bash
+#!/bin/bash
+./calculate < DataBundle > result 2> errors
+if [ -s errors]
+then
+     ./verify < result
+else
+     echo "error occured during execution."
+fi
+```
+Another useful command is `|`, which chained linux commands together, as such the output of previous commands is passed as input to the next command.
+
+Suppose you want to check the status of process `p` running on your system. Thus you typed in `ps aux`, and all the processes are listed, but there are too many processes that you can't find `p`. As such, you can pass the result of `ps aux` to `grep 'p'` to capture process `p`'s status. Thus the command will be look like `ps aux | grep 'p'`.
+
+Input/output streams can easily be redirected, allowing information to be sent or received from files or other applications(commands and user programs). This can mean test data can be easily supplied or output captured.
+
+### Benefit: Defined Standards for Help
+If someone does not know a how to use certain command, they can easily see what the it provides with `man` command. There are also defined wildcard syntax standards for specifying multiple files, drawing on likely existing developer knowledge.
+
+For example, if you are new to bash and want to learn a new command `grep`, you can simply type in `man grep` in the terminal and a help page will show up. It thoroughly include **What**, **How** as well as many other details of the command.
+
+```
+GREP(1)                   BSD General Commands Manual                  GREP(1)
+
+NAME
+     grep, egrep, fgrep, zgrep, zegrep, zfgrep -- file pattern searcher
+
+SYNOPSIS
+     grep [-abcdDEFGHhIiJLlmnOopqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
+          [-e pattern] [-f file] [--binary-files=value] [--color[=when]]
+          [--colour[=when]] [--context[=num]] [--label] [--line-buffered]
+          [--null] [pattern] [file ...]
+
+DESCRIPTION
+     The grep utility searches any given input files, selecting lines that
+     match one or more patterns.  By default, a pattern matches an input line
+     if the regular expression (RE) in the pattern matches the input line
+     without its trailing newline.  An empty expression matches every line.
+     Each input line that matches at least one of the patterns is written to
+     the standard output ......
+
+```
+## How to Get Started with linux command-line based shell?
```

Comment:
This part here just feels like your saying, 'jump right in and start using it' !

I liked your git example here! Maybe you can elaborate a bit more on specific use cases when I should turn to the command line?

Reply:
> I originally thought that this article would be on learning how to use the linux shell, but it's really just why you should start learning.`
That's because the requirement of book chapter is to focus more on what and why.